src/GafferScene/Instancer.cpp

```diff
@@ -857,10 +1497,127 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
 		static_cast<AtomicCompoundDataPlug *>( output )->setValue( result );
 		return;
 	}
+	else if( output == variationsPlug() )
+	{
+		// Compute the number of variations by accumulating massive lists of unique hashes from all EngineDatas
+		// and then counting the total number of uniques
+
+		tbb::spin_mutex locationMutex;
+		std::vector< std::shared_ptr< EngineData::PrototypeHashes > > perLocationHashes;
+		std::shared_ptr< std::vector< InternedString > > contextVariableNames;
+		struct HashSetAccum
+		{
+			HashSetAccum(
+				const ObjectPlug *enginePlug,
+				tbb::spin_mutex &locationMutex,
+				std::vector< std::shared_ptr< EngineData::PrototypeHashes > > &perLocationHashes,
+				std::shared_ptr< std::vector< InternedString > > &contextVariableNames
+			):
+				m_enginePlug( enginePlug ), m_locationMutex( locationMutex ),
+				m_perLocationHashes( perLocationHashes ), m_contextVariableNames( contextVariableNames )
+			{
+			}
+			bool operator()( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+			{
+				ConstEngineDataPtr engine = boost::static_pointer_cast<const EngineData>( m_enginePlug->getValue() );
+				std::shared_ptr< EngineData::PrototypeHashes > locationHashes = engine->uniquePrototypeHashes();
+
+				tbb::spin_mutex::scoped_lock lock( m_locationMutex );
+				m_perLocationHashes.push_back( locationHashes );
+				if( !m_contextVariableNames )
+				{
+					// The context variable names are guaranteed to match for all EngineData, so we just
+					// grab them from the first EngineData we process
+					m_contextVariableNames = std::shared_ptr< std::vector< InternedString > >( new std::vector< InternedString >( engine->contextVariableNames() ) );
+				}
+				return true;
+			}
+		private:
+			const ObjectPlug *m_enginePlug;
+			tbb::spin_mutex &m_locationMutex;
+			std::vector< std::shared_ptr< EngineData::PrototypeHashes > > &m_perLocationHashes;
+			std::shared_ptr< std::vector< InternedString > > &m_contextVariableNames;
```

Comment:
Could we get rid of all this boilerplate by passing a lambda to `filteredParallelTraverse()`? If so, I think it would be a lot more readable.

Reply:
Switched.